pep-0689.rst

```diff
@@ -0,0 +1,231 @@
+defined. In this case, using them will generate a deprecation warning on
+compilers that support ``Py_DEPRECATED``.
+
+A similar deprecation period will be used when making more APIs semi-stable
+in the future:
+
+- When moving from public API, the deprecation period should follow Python's
+  backwards compatibility policy (currently, it should last at least
+  two releases).
+- When moving from API that is documented as unstable,
+  the deprecation period can only last one release.
+- When moving from private API or adding new API, no deprecation period
+  is necessary.
+
+Leading underscores will be removed from the names of the moved APIs.
+The old underscored name of a renamed API will be available (as an alias
+using ``#define``) at least until that API changes.
+
+The semi-stable C-API tier and ``Py_USING_SEMI_STABLE_API`` will be documented,
+and documentation of each semi-stable API will be updated.
+
+
+Adjustments during Beta periods
+-------------------------------
+
+New APIs can be added to the semi-stable tier, and private APIs can be moved
+to it, up to the first release candidate of a new minor version.
+Consensus on the ``capi-sig`` or ``python-dev`` is needed it the beta period.
+
+In the Beta period, no API may be moved to more private tier, e.g.
+what is public in Beta 1 must stay public until the final release.
+
+
+Initial semi-stable API
+-----------------------
+
+The following API will initially be semi-stable.
+The set may be adjusted for 3.11.
+
+Code object constructors:
+
+- ``PyCode_New()``
+- ``PyCode_NewWithPosOnlyArgs()``
+
+Frame evaluation API (PEP 523):
+
+- ``_PyFrameEvalFunction``
+- ``_PyInterpreterState_GetEvalFrameFunc()``
+- ``_PyInterpreterState_SetEvalFrameFunc()``
+- ``_PyEval_RequestCodeExtraIndex()``
+- ``_PyCode_GetExtra()``
+- ``_PyCode_SetExtra()``
+- ``struct _PyInterpreterFrame`` (as an incomplete, opaque struct)
+- ``_PyFrame_GetFrameObject``
+- ``PyEval_EvalFrameDefault``
+  (new function that calls ``_PyEval_EvalFrameDefault``, but takes
+  ``PyFrameObject`` rather than ``_PyInterpreterFrame``)
+
+(Leading underscores will be removed as mentioned above.)
+
+
+Backwards Compatibility
+=======================
+
+The C API backwards compatibility story will be made clearer.
+
+
+Security Implications
+=====================
+
+None known.
+
+
+How to Teach This
+=================
+
+The changes affect advanced C programmers, who should consult the
+updated reference documentation, devguide and/or What's New document·.
+
+
+Reference Implementation
+========================
+
+https://github.com/python/cpython/issues/91744
+
+
+Rejected Ideas
+==============
+
+It might be good to add a similar tier in the Python (not C) API,
+e.g. for ``types.CodeType``.
+However, the opt-in mechanism would need to be different (if any).
+This is outside the scope of the PEP.
+
+
+Open Issues
+===========
+
+- “Semi-stable” is not a perfect name.
```

Comment:
I don't really like the name either, but haven't been able to think of anything better.